File: scripts/_schema.py
"""Tiny jsonschema validator subset for offline use."""
from __future__ import annotations

from typing import Any, Dict


class ValidationError(Exception):
    pass


def _assert(condition: bool, message: str) -> None:
    if not condition:
        raise ValidationError(message)


def _validate(instance: Any, schema: Dict[str, Any], path: str) -> None:
    schema_type = schema.get("type")
    if schema_type == "object" or (schema_type is None and "properties" in schema):
        _assert(isinstance(instance, dict), f"{path or '<root>'} must be an object")
        required = schema.get("required", [])
        for field in required:
            _assert(field in instance, f"Missing required field {path + '.' if path else ''}{field}")
        properties = schema.get("properties", {})
        for key, value in instance.items():
            if key in properties:
                _validate(value, properties[key], f"{path}.{key}" if path else key)
        if not schema.get("additionalProperties", True):
            for key in instance:
                _assert(key in properties, f"Unexpected field {path}.{key}" if path else key)
        return
    if schema_type == "array":
        _assert(isinstance(instance, list), f"{path} must be an array")
        min_items = schema.get("minItems")
        max_items = schema.get("maxItems")
        if min_items is not None:
            _assert(len(instance) >= min_items, f"{path} must have at least {min_items} items")
        if max_items is not None:
            _assert(len(instance) <= max_items, f"{path} must have at most {max_items} items")
        item_schema = schema.get("items")
        if item_schema:
            for idx, item in enumerate(instance):
                _validate(item, item_schema, f"{path}[{idx}]")
        return
    if schema_type == "boolean":
        _assert(isinstance(instance, bool), f"{path} must be a boolean")
        return
    if schema_type == "number":
        _assert(isinstance(instance, (int, float)), f"{path} must be a number")
        return
    if schema_type == "string":
        _assert(isinstance(instance, str), f"{path} must be a string")
        if "enum" in schema:
            _assert(instance in schema["enum"], f"{path} must be one of {schema['enum']}")
        return
    if "enum" in schema:
        _assert(instance in schema["enum"], f"{path} must be one of {schema['enum']}")
        return
    # Fallback: accept any type


def validate(instance: Any, schema: Dict[str, Any]) -> None:
    _validate(instance, schema, "")


__all__ = ["validate", "ValidationError"]

File: scripts/_schema.ts
// RESUME_MARKER: CHK_2
/* eslint-disable @typescript-eslint/no-explicit-any */

export class ValidationError extends Error {}

function assertCondition(condition: boolean, message: string): void {
  if (!condition) {
    throw new ValidationError(message);
  }
}

function formatPath(path: string): string {
  return path || '<root>';
}

function validateInternal(instance: any, schema: Record<string, any>, path: string): void {
  const schemaType = schema.type;
  if (schemaType === 'object' || (schemaType == null && schema.properties)) {
    assertCondition(isPlainObject(instance), `${path || '<root>'} must be an object`);
    const required: string[] = Array.isArray(schema.required) ? schema.required : [];
    for (const field of required) {
      const fieldPath = path ? `${path}.${field}` : field;
      assertCondition(Object.prototype.hasOwnProperty.call(instance, field), `Missing required field ${fieldPath}`);
    }
    const properties: Record<string, any> = schema.properties || {};
    for (const [key, value] of Object.entries(instance)) {
      if (properties && Object.prototype.hasOwnProperty.call(properties, key)) {
        const childPath = path ? `${path}.${key}` : key;
        validateInternal(value, properties[key], childPath);
      }
    }
    if (schema.additionalProperties === false) {
      for (const key of Object.keys(instance)) {
        if (!Object.prototype.hasOwnProperty.call(properties, key)) {
          const fieldPath = path ? `${path}.${key}` : key;
          assertCondition(false, `Unexpected field ${fieldPath}`);
        }
      }
    }
    return;
  }
  if (schemaType === 'array') {
    assertCondition(Array.isArray(instance), `${formatPath(path)} must be an array`);
    if (typeof schema.minItems === 'number') {
      assertCondition(
        instance.length >= schema.minItems,
        `${formatPath(path)} must have at least ${schema.minItems} items`
      );
    }
    if (typeof schema.maxItems === 'number') {
      assertCondition(
        instance.length <= schema.maxItems,
        `${formatPath(path)} must have at most ${schema.maxItems} items`
      );
    }
    if (schema.items) {
      instance.forEach((item: any, idx: number) => {
        validateInternal(item, schema.items, `${path}[${idx}]`);
      });
    }
    return;
  }
  if (schemaType === 'boolean') {
    assertCondition(typeof instance === 'boolean', `${formatPath(path)} must be a boolean`);
    return;
  }
  if (schemaType === 'number') {
    assertCondition(typeof instance === 'number', `${formatPath(path)} must be a number`);
    return;
  }
  if (schemaType === 'string') {
    assertCondition(typeof instance === 'string', `${formatPath(path)} must be a string`);
    if (Array.isArray(schema.enum)) {
      assertCondition(schema.enum.includes(instance), `${formatPath(path)} must be one of ${schema.enum}`);
    }
    return;
  }
  if (Array.isArray(schema.enum)) {
    assertCondition(schema.enum.includes(instance), `${formatPath(path)} must be one of ${schema.enum}`);
    return;
  }
}

function isPlainObject(value: unknown): value is Record<string, any> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

export function validate(instance: any, schema: Record<string, any>): void {
  validateInternal(instance, schema, '');
}

export default { validate, ValidationError };
